roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cpp

```diff
@@ -525,5 +512,92 @@ TEST_P(LikelihoodGradientJob, Gaussian1DAlsoWithLikelihoodJob)
    EXPECT_FLOAT_EQ(minNll0, minNll1);
    EXPECT_FLOAT_EQ(mu0, mu1);
    EXPECT_FLOAT_EQ(muerr0, muerr1);
-   EXPECT_NEAR_REL(edm0, edm1, 1e-5);
+   EXPECT_NEAR(edm0, edm1, 1e-5);
+}
+
+TEST_F(LikelihoodSimBinnedConstrainedTest, ConstrainedAndOffsetAlsoWithLikelihoodJob)
+{
+   // do a minimization, but now using GradMinimizer and its MP version
+   nll.reset(pdf->createNLL(*data, RooFit::Constrain(RooArgSet(*w.var("alpha_bkg_obs_A"))),
+                            RooFit::GlobalObservables(RooArgSet(*w.var("alpha_bkg_obs_B"))), RooFit::Offset(true)));
+
+   // parameters
+   std::size_t NWorkers = 2;
+
+   RooArgSet *values = pdf->getParameters(data);
+
+   values->add(*pdf);
+   values->add(*nll);
+
+   RooArgSet *savedValues = dynamic_cast<RooArgSet *>(values->snapshot());
+   if (savedValues == nullptr) {
+      throw std::runtime_error("params->snapshot() cannot be casted to RooArgSet!");
+   }
```

Comment:
```suggestion
   RooArgSet values;
   pdf->getParameters(data.get(), values);

   values.add(*pdf);
   values.add(*nll);
```
There are two memory leaks here: from `getParameters`, and from `snapshot.

It would be better to use the stack version of `getParameters`, and then you also don't need to do the `dynamic_cast` because the output is a RooArgSet anyway.

These leaks also appear in the other unit tests in this file and it would be good to fix them too, could be even in the same commit.

Reply:
Wow, you're right, good catch. I think I left the `snapshot`s like this when I once tried to convert them to `unique_ptr`s, but that didn't work with the `*values = *savedValues` lines, so then I just forgot to do it properly some other way.

I have implemented your suggestions.